### PR TITLE
java_bytecode tweaks

### PR DESCRIPTION
**Pull request by LAJW**
- [x] Each commit message has a non-empty body, explaining why the change was made.
- [n/a] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [n/a] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [n/a] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [n/a] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own. (where possible)

### CODE REVIEW

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp [hunk 1]
```diff
@@ -38,7 +38,7 @@ class java_bytecode_parsert:public parsert
   {
   }
 
-  virtual bool parse();
+  virtual bool parse() override;
```
Comment (owen-mc-diffblue) [inline jbmc/src/java_bytecode/java_bytecode_parser.cpp]: :pick: The coding standard says you should remove the `virtual` since you have `override`.

**Review by owen-mc-diffblue — APPROVED** [review]

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp (hunk 1 above)
Comment (owen-mc-diffblue) [reply jbmc/src/java_bytecode/java_bytecode_parser.cpp]: Indeed, the linter is failing for this reason

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp [hunk 2]
```diff
@@ -137,8 +137,12 @@ class java_bytecode_parsert:public parsert
     }
   }
 
-  u8 read_bytes(size_t bytes)
+  template <typename T>
```
Comment (smowton) [inline jbmc/src/java_bytecode/java_bytecode_parser.cpp:131]: commit message "Make read_bytes less likely to be called incorrectly" truncated (larger than... what?)

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp [hunk 3]
```diff
@@ -653,15 +653,9 @@ void java_bytecode_parsert::rconstant_pool()
 
   constant_pool.resize(constant_pool_count);
 
-  for(constant_poolt::iterator
-      it=constant_pool.begin();
-      it!=constant_pool.end();
+  for(auto it = std::next(constant_pool.begin()); it != constant_pool.end();
```
Comment (smowton) [inline jbmc/src/java_bytecode/java_bytecode_parser.cpp:656]: Could the constant pool be empty? If so this is wrong

**Comment by smowton** [comment]
Looks ok except the `std::next` on a maybe-end iterator, and I don't see the value in inlining e.g. `annotationst`, where I guess the idea was if we want to turn that into `std::list` or whatever then we only need to make the change in one place?

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp (hunk 3 above)
Comment (LAJW) [reply jbmc/src/java_bytecode/java_bytecode_parser.cpp:656]: No, this is already checked at the start of the function.

**Comment by LAJW** [comment]
@smowton I don't like obstructing collections like that, because each type has different properties and advantages. Code already relies on these assumptions, so swapping containers is not really an option. Having it spelled out in full makes it easier to see what you can/cannot use. I've seen cases where people were manually converting vectors, just because they were typedefed.

And in this case this code was just stealing typedefs from other unrelated class for convenience, not because this is what a particular class returns.

**Comment by smowton** [comment]
Eh, if you want. It's widely done in this codebase so you'll have much more to do if you want to go that way, and hundreds of lines of churn for your trouble.

**Comment by LAJW** [comment]
@smowton Commit message updated. @owen-jones-diffblue linter problems fixed.

**Review by smowton — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 8d3559bba095fb725ad6fa7a7b45a5880e56b155).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/120884008

**Pull request merged**